### PR TITLE
Add Designed for iPad attach destination for ARM macOS

### PR DESCRIPTION
**Pull request by jmagman**
Add a new `MacOSDesignedForIPadDevice` device to target when running [native ARM iOS apps on Apple Silicon macOS](https://developer.apple.com/documentation/apple-silicon/running-your-ios-apps-on-macos).  I called this "Mac Designed for iPad" based on the Xcode target name, since there doesn't seem to be a pithy marketing name for this feature.
Screen Shot 2021-06-10 at 10 22 49 PM

It only supports `flutter attach`ing to iOS apps already launched from Xcode, per guidance from product leadership so as not to confuse the desktop macOS development story.

The running apps are discovered via mDNS.
Screen Shot 2021-06-10 at 9 36 26 PM

It's like a desktop device for log reading/port purposes.  It's like an iOS device in that it's only supported for apps with an `ios` project directory.  If `startApp`, `buildApp`, etc are ever implemented in a future PR, it will have to create the right structured bundle out of the iOS app. 
Screen Shot 2021-06-10 at 10 26 36 PM

Also create `MacOSDesignedForIPadDevices` device discovery that will only return this new device type if on an ARM Mac, and iOS is supported (Xcode is installed, etc).

`flutter devices` does not show this device type:
```
$ flutter devices
No devices detected.
```
However when I `flutter attach`:
```
$ flutter attach --app-id "com.example.testCreate"
Syncing files to device Mac Designed for iPad...                    5.9s

Flutter run key commands.
r Hot reload. 🔥🔥🔥
...
```
And it's actually attached:
https://user-images.githubusercontent.com/682784/121634738-f78b9a00-ca39-11eb-9a25-b53f16dd2cda.mov

Will try to figure out a way to integration test this once https://github.com/flutter/flutter/issues/79430 is done.

`flutter attach` part of https://github.com/flutter/flutter/issues/66184

### CODE REVIEW

File: packages/flutter_tools/lib/src/macos/macos_ipad_device.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  bool get supportsPlatform => _platform.isMacOS;
+
+  /// iOS (not desktop macOS) development is enabled, the host is an ARM Mac,
+  /// and discovery is allowed for this command.
+  @override
+  bool get canListAnything =>
+      _iosWorkflow.canListDevices && _operatingSystemUtils.hostPlatform == HostPlatform.darwin_arm && allowDiscovery;
+
+  /// Set to show ARM macOS as an iOS device target.
+  static bool allowDiscovery = false;
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/macos/macos_ipad_device.dart:128]: I couldn't figure out a non-global way to set this (modeled on `FlutterTesterDevices.showFlutterTesterDevice`).  I tried to make it a property on a context override on `IOSWorkflow` but I would have needed to set it the context override only for the `attach` command in: 
https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter_tools/lib/src/runner/flutter_command_runner.dart#L232-L235
and
https://github.com/flutter/flutter/blob/80668d20f86be0fefd9f210db0d23c4bb2d68d59/packages/flutter_tools/lib/src/commands/attach.dart#L178-L179
and
https://github.com/flutter/flutter/blob/80668d20f86be0fefd9f210db0d23c4bb2d68d59/packages/flutter_tools/lib/src/commands/attach.dart#L210-L213

Welcome suggestions to make this better.

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart [hunk 2]
```diff
@@ -58,6 +59,10 @@ final vm_service.Isolate fakeUnpausedIsolate = vm_service.Isolate(
 );
 
 void main() {
+  tearDown(() {
+    MacOSDesignedForIPadDevices.allowDiscovery = false;
+  });
```
Comment (jmagman) [inline packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:64]: See `allowDiscovery` global comment.  It's similar to:
https://github.com/flutter/flutter/blob/6beafa7b16064a78a496b2b400e31920ca297c08/packages/flutter_tools/test/general.shard/tester/flutter_tester_test.dart#L43-L46

**Comment by jonahwilliams** [comment]
Heads up that I won't have a chance to look at this today since I'm heading out early, but I will take a look first thing monday

File: packages/flutter_tools/lib/src/macos/macos_ipad_device.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/macos/macos_ipad_device.dart:128]: static is fine, please no more context overrides

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by thipokch** [comment]
@jmagman Just curious if attaching from vscode would work too?

**Comment by thipokch** [comment]
```json 
// (launch.json)
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Attach: Design for iPad",
            "request": "attach",
            "type": "dart",
            "program": "lib/main.dart",
            "args": [
                "--app-id",
                "com.example"
            ],
            "deviceId": "designed-for-ipad",
        }
    ]
}
```

**Pull request merged**